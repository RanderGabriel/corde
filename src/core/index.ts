import { CordeBot } from "./cordeBot";

export { CordeBot } from "./cordeBot";
