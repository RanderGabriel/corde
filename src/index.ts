/**
 * @file Automatically generated by barrelsby.
 */

 export * from "./cli";
export * from "./erros/index";
