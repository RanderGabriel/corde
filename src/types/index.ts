export * from "./types";
export * from "./locale";
export * from "./discordTypes";
