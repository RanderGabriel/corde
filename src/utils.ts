import { ParameterNotFoundError } from './errors';

export function clone(object: object | object[]): object | object[] {
  if (!object) {
    throw new ParameterNotFoundError('Object can not be undefined');
  }

  if (object as object[]) {
    const copyArray = [];
    for (const i in object as []) {
      copyArray.push(copyObject((object as [])[i]));
    }
    return copyArray;
  } else {
    return copyObject(object);
  }
}

function copyObject(object: object) {
  return JSON.parse(JSON.stringify(object));
}
