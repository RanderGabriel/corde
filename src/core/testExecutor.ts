import chalk from "chalk";
import { runtime } from "../common/runtime";
import { testCollector } from "../common/testCollector";
import { printHookErrors } from "../common/printHookError";
import {
  MESSAGE_TAB_SPACE,
  TAG_FAIL,
  TAG_PASS,
  TAG_PENDING,
  TEST_FAIL_ICON,
  TEST_PASSED_ICON,
  TEST_RUNNING_ICON,
  TEXT_EMPTY,
  TEXT_FAIL,
  TEXT_PASS,
} from "../consts";
import { Queue } from "../data-structures";
import {
  Group,
  RunnerReport,
  SemiRunnerReport,
  Test,
  TestFile,
  TestReport,
  VoidLikeFunction,
} from "../types";
import { formatObject, stringIsNullOrEmpty, Timer } from "../utils";
import { LogUpdate } from "../utils";

export class TestExecutor {
  private _logUpdate: LogUpdate;

  constructor(logUpdate: LogUpdate) {
    this._logUpdate = logUpdate;
  }

  async runTestsAndPrint(testFiles: TestFile[]): Promise<RunnerReport> {
    const testsTimer = new Timer();
    testsTimer.start();

    const semiReport: SemiRunnerReport = {
      totalTests: 0,
      totalTestFiles: 0,
      totalTestsPassed: 0,
      totalTestsFailed: 0,
      totalTestFilesPassed: 0,
      totalTestFilesFailed: 0,
      totalEmptyTestFiles: 0,
      totalEmptyTests: 0,
    };

    for (const testFile of testFiles) {
      if (testFile.isEmpty) {
        this._logUpdate.append(`${TAG_PENDING("EMPTY")}  ${testFile.path}`);
        this._logUpdate.persist();
        semiReport.totalEmptyTestFiles++;
        semiReport.totalTestFiles++;
      } else {
        if (await this.executeTestFile(testFile, semiReport)) {
          semiReport.totalTestFilesPassed++;
        } else {
          semiReport.totalTestFilesFailed++;
        }
      }
    }
    const testsDiff = testsTimer.stop();

    return {
      testTimer: testsDiff[0],
      ...semiReport,
    };
  }

  private async executeTestFile(testFile: TestFile, semiReport: SemiRunnerReport) {
    const testFileTimer = new Timer();
    semiReport.totalTestFiles++;
    let fileHasPassed = true;

    testFileTimer.start();
    const logIndex = this._logUpdate.append(`${TAG_PENDING()}  ${testFile.path}`);

    for (const group of testFile.groups) {
      if (!(await this.executeGroup(group, semiReport))) {
        fileHasPassed = false;
      }
    }

    const _diff = testFileTimer.stop();

    let fileLabel = TAG_PASS();
    let fileNameLabel = testFile.path;

    if (!fileHasPassed) {
      fileLabel = TAG_FAIL();
      fileNameLabel = chalk.red(fileNameLabel);
    }

    this._logUpdate.updateLine(
      logIndex,
      `${fileLabel}  ${fileNameLabel}   ${chalk.cyan(_diff[0])}`,
    );

    this._logUpdate.persist();

    return fileHasPassed;
  }

  private async executeGroup(group: Group, semiReport: SemiRunnerReport) {
    const tests = this.getAssertionPropsFromGroup(group);
    let fileHasPassed = true;
    for (const test of tests) {
      if (!(await this.executeTest(test, semiReport, group.name))) {
        fileHasPassed = false;
      }
    }
    return fileHasPassed;
  }

  private async executeTest(
    test: Test,
    semiReport: SemiRunnerReport,
    groupName?: string | number | boolean,
  ) {
    const testTimer = new Timer();
    let fileHasPassed = true;
    testTimer.start();
    let logPosition = 0;

    const testText = this.createTestText(test.name);
    logPosition = this._logUpdate.appendLine(testText);

    const reports = await this.runTest(test);

    if (reports.length === 0) {
      semiReport.totalEmptyTests++;
      semiReport.totalTests++;
    }

    const testDiff = testTimer.stop();

    const testNameLabel = this.testReportLabelFunction(reports);

    const formatedGroupName = !stringIsNullOrEmpty(groupName) ? groupName + " -> " : "";

    if (stringIsNullOrEmpty(test.name)) {
      this._logUpdate.updateLine(
        logPosition,
        `${MESSAGE_TAB_SPACE}${testNameLabel(
          formatedGroupName + "<empty test name>",
        )}   ${chalk.cyan(testDiff[0])}`,
      );
    } else {
      this._logUpdate.updateLine(
        logPosition,
        `${MESSAGE_TAB_SPACE}${testNameLabel(formatedGroupName + test.name)}   ${chalk.cyan(
          testDiff[0],
        )}`,
      );
    }

    if (!this.printTestsReportAndUpdateRunnerReport(reports, semiReport)) {
      fileHasPassed = false;
    }

    test.testsReports = reports;
    return fileHasPassed;
  }

  private printTestsReportAndUpdateRunnerReport(
    reports: TestReport[],
    semiReport: SemiRunnerReport,
  ) {
    let fileHasPassed = true;
    for (const report of reports) {
      semiReport.totalTests++;
      if (!this.printReportAndUpdateRunnerReport(report, semiReport)) {
        fileHasPassed = false;
      }
    }
    return fileHasPassed;
  }

  private testReportLabelFunction(reports: TestReport[]) {
    if (reports.length === 0) {
      return (text: string) => TEXT_EMPTY(" " + TEST_RUNNING_ICON + " " + text + " (empty)");
    }

    if (reports.some((report) => !report.pass)) {
      return (text: string) => TEXT_FAIL(TEST_FAIL_ICON + " " + text);
    }
    return (text: string) => TEXT_PASS(TEST_PASSED_ICON + " " + text);
  }

  private printReportAndUpdateRunnerReport(report: TestReport, semiReport: SemiRunnerReport) {
    if (report.pass) {
      semiReport.totalTestsPassed++;
      return true;
    }

    semiReport.totalTestsFailed++;
    this.printReportData(report);

    return false;
  }

  private printReportData(report: TestReport) {
    if (report.message) {
      this._logUpdate.appendLine(report.message);
    }

    if (!report.pass && report.trace) {
      this._logUpdate.appendLine(report.trace);
    }
  }

  private createTestText(testName: string | number | boolean) {
    if (stringIsNullOrEmpty(testName)) {
      return this.createTestTextByStatus("<empty test name>");
    }
    return this.createTestTextByStatus(testName);
  }

  private createTestTextByStatus(testName: string | number | boolean) {
    const icon = TEST_RUNNING_ICON;
    return `${MESSAGE_TAB_SPACE}${icon} ${testName}`;
  }

  async runTest(test: Test) {
    const reports: TestReport[] = [];

    // before e after hooks will run just one time
    // for test structure (if the hook fail)
    let keepRunningBeforeEachFunctions = true;
    let keepRunningAfterEachFunctions = true;

    for (const testfn of test.testsFunctions) {
      keepRunningBeforeEachFunctions = await this.executeHookFunctionIfPossible(
        keepRunningBeforeEachFunctions,
        testCollector.beforeEachFunctions,
      );

      let _report: TestReport;
      try {
        _report = await runtime.injectBot(testfn);
      } catch (error) {
        _report = {
          pass: false,
          message: this.getErrorMessage(error),
        };
      }

      keepRunningAfterEachFunctions = await this.executeHookFunctionIfPossible(
        keepRunningAfterEachFunctions,
        testCollector.afterEachFunctions,
      );

      reports.push(_report);
    }
    return reports;
  }

  private async executeHookFunctionIfPossible(
    keepRunning: boolean,
    queues: Queue<VoidLikeFunction>,
  ) {
    if (keepRunning) {
      const _functionErrors = await queues.executeWithCatchCollectAsync();
      if (_functionErrors && _functionErrors.length) {
        printHookErrors(_functionErrors);
        return false;
      }
      return true;
    }
    return false;
  }

  private getErrorMessage(error: any) {
    if (error instanceof Error) {
      return error.message;
    }

    return formatObject(error);
  }

  private getAssertionPropsFromGroup(group: Group) {
    const assertions: Test[] = [];
    if (group.tests) {
      group.tests.forEach((test) => {
        assertions.push(...this.getAssertionsPropsFromTest(test));
      });
    }

    if (group.subGroups) {
      group.subGroups.forEach((subGroup) => {
        const subGroupTests = this.getAssertionPropsFromGroup(subGroup);
        assertions.push(...subGroupTests);
      });
    }

    return assertions;
  }

  private getAssertionsPropsFromTest(test: Test) {
    const tests: Test[] = [test];

    if (test.subTests) {
      test.subTests.forEach((subtest) => {
        tests.push(...this.getAssertionsPropsFromTest(subtest));
      });
    }

    return tests;
  }
}
