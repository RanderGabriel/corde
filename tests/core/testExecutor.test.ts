import { TestExecutor } from "../../src/core/testExecutor";
import { SemiRunnerReport, TestFile } from "../../src/types";
import { buildReportMessage, diff, LogUpdate } from "../../src/utils";
import {
  generateTestFile,
  removeANSIColorStyle,
  TestFileGeneratorInfo,
  testFileNames,
  testNames,
} from "../testHelper";

import { mockTimer } from "../mocks/mockTimer";

let logUpdate: LogUpdate;
let testRunner: TestExecutor;

mockTimer();

beforeEach(() => {
  logUpdate = new LogUpdate();
  testRunner = new TestExecutor(logUpdate);
});

it("should print report for 1 test file, 1 test clausure and 1 test function", async () => {
  const data: TestFileGeneratorInfo = {
    amountOfTestFiles: 1,
    amountOfTestFunctions: 1,
    amountOfTests: 1,
  };

  const tests = generateTestFile(data);

  const report = await testRunner.runTestsAndPrint(tests);
  const amountOfTest = data.amountOfTestFiles * data.amountOfTests * data.amountOfTestFunctions;

  expect(report).toMatchObject<SemiRunnerReport>({
    totalTestFiles: data.amountOfTestFiles,
    totalTestFilesFailed: 0,
    totalTestFilesPassed: data.amountOfTestFiles,
    totalTests: amountOfTest,
    totalTestsFailed: 0,
    totalTestsPassed: amountOfTest,
    totalEmptyTestFiles: 0,
    totalEmptyTests: 0,
  });
  expect(removeANSIColorStyle(logUpdate.stder)).toMatchSnapshot();
});

it("should print report for 1 test file, 1 test clausure and 2 test function", async () => {
  const data: TestFileGeneratorInfo = {
    amountOfTestFiles: 1,
    amountOfTestFunctions: 2,
    amountOfTests: 1,
  };

  const tests = generateTestFile(data);

  const report = await testRunner.runTestsAndPrint(tests);
  const amountOfTest = data.amountOfTestFiles * data.amountOfTests * data.amountOfTestFunctions;

  expect(report).toMatchObject<SemiRunnerReport>({
    totalTestFiles: data.amountOfTestFiles,
    totalTestFilesFailed: 0,
    totalTestFilesPassed: data.amountOfTestFiles,
    totalTests: amountOfTest,
    totalTestsFailed: 0,
    totalTestsPassed: amountOfTest,
    totalEmptyTestFiles: 0,
    totalEmptyTests: 0,
  });
  expect(removeANSIColorStyle(logUpdate.stder)).toMatchSnapshot();
});

it("should print report for 2 test file, 1 test clausure and 1 test function", async () => {
  const data: TestFileGeneratorInfo = {
    amountOfTestFiles: 2,
    amountOfTestFunctions: 1,
    amountOfTests: 1,
  };

  const tests = generateTestFile({
    amountOfTestFiles: 2,
    amountOfTestFunctions: 1,
    amountOfTests: 1,
  });

  const report = await testRunner.runTestsAndPrint(tests);
  const amountOfTest = data.amountOfTestFiles * data.amountOfTests * data.amountOfTestFunctions;

  expect(report).toMatchObject<SemiRunnerReport>({
    totalTestFiles: data.amountOfTestFiles,
    totalTestFilesFailed: 0,
    totalTestFilesPassed: data.amountOfTestFiles,
    totalTests: amountOfTest,
    totalTestsFailed: 0,
    totalTestsPassed: amountOfTest,
    totalEmptyTestFiles: 0,
    totalEmptyTests: 0,
  });
  expect(removeANSIColorStyle(logUpdate.stder)).toMatchSnapshot();
});

it("should print report for 2 test file, 1 test clausure and 2 test function", async () => {
  const data: TestFileGeneratorInfo = {
    amountOfTestFiles: 2,
    amountOfTestFunctions: 2,
    amountOfTests: 1,
  };

  const tests = generateTestFile(data);

  const report = await testRunner.runTestsAndPrint(tests);
  const amountOfTest = data.amountOfTestFiles * data.amountOfTests * data.amountOfTestFunctions;

  expect(report).toMatchObject<SemiRunnerReport>({
    totalTestFiles: data.amountOfTestFiles,
    totalTestFilesFailed: 0,
    totalTestFilesPassed: data.amountOfTestFiles,
    totalTests: amountOfTest,
    totalTestsFailed: 0,
    totalTestsPassed: amountOfTest,
    totalEmptyTestFiles: 0,
    totalEmptyTests: 0,
  });
  expect(removeANSIColorStyle(logUpdate.stder)).toMatchSnapshot();
});

it("should print report for 2 test file, 2 test clausure and 2 test function", async () => {
  const data: TestFileGeneratorInfo = {
    amountOfTestFiles: 2,
    amountOfTestFunctions: 2,
    amountOfTests: 2,
  };

  const tests = generateTestFile(data);

  const report = await testRunner.runTestsAndPrint(tests);
  const amountOfTest = data.amountOfTestFiles * data.amountOfTests * data.amountOfTestFunctions;

  expect(report).toMatchObject<SemiRunnerReport>({
    totalTestFiles: data.amountOfTestFiles,
    totalTestFilesFailed: 0,
    totalTestFilesPassed: data.amountOfTestFiles,
    totalTests: amountOfTest,
    totalTestsFailed: 0,
    totalTestsPassed: amountOfTest,
    totalEmptyTestFiles: 0,
    totalEmptyTests: 0,
  });
  expect(removeANSIColorStyle(logUpdate.stder)).toMatchSnapshot();
});

it("should print report for 1 test file, 1 test clausure and 1 test function that fail", async () => {
  const data: TestFileGeneratorInfo = {
    amountOfTestFiles: 1,
    testFunctionsReport: [
      {
        pass: false,
        message: buildReportMessage("expected: hi\n", "received: hi!"),
      },
    ],
    amountOfTests: 1,
  };

  const tests = generateTestFile(data);

  const report = await testRunner.runTestsAndPrint(tests);
  const amountOfTest = data.amountOfTestFiles * data.amountOfTests * data.amountOfTestFunctions;

  expect(report).toMatchObject<SemiRunnerReport>({
    totalTestFiles: data.amountOfTestFiles,
    totalTestFilesFailed: 1,
    totalTestFilesPassed: 0,
    totalTests: amountOfTest,
    totalTestsFailed: amountOfTest,
    totalTestsPassed: 0,
    totalEmptyTestFiles: 0,
    totalEmptyTests: 0,
  });
  expect(removeANSIColorStyle(logUpdate.stder)).toMatchSnapshot();
});

it("should print report for 1 test file, 1 test clausure and 1 failed function and 1 passed function", async () => {
  const data: TestFileGeneratorInfo = {
    amountOfTestFiles: 1,
    testFunctionsReport: [
      {
        pass: false,
        message: buildReportMessage("expected: hi\n", "received: hi!"),
      },
      { pass: true },
    ],
    amountOfTests: 1,
  };

  const tests = generateTestFile(data);

  const report = await testRunner.runTestsAndPrint(tests);

  expect(report).toMatchObject<SemiRunnerReport>({
    totalTestFiles: data.amountOfTestFiles,
    totalTestFilesFailed: 1,
    totalTestFilesPassed: 0,
    totalTests: 2,
    totalTestsFailed: 1,
    totalTestsPassed: 1,
    totalEmptyTestFiles: 0,
    totalEmptyTests: 0,
  });
  expect(removeANSIColorStyle(logUpdate.stder)).toMatchSnapshot();
});

it("should print report for 1 test file, 1 test clausure and 1 failed function and 1 passed function", async () => {
  const data: TestFileGeneratorInfo = {
    amountOfTestFiles: 1,
    testFunctionsReport: [
      {
        pass: false,
        message: buildReportMessage("expected: hi\n", "received: hi!"),
      },
      { pass: true },
    ],
    amountOfTests: 1,
  };

  const tests = generateTestFile(data);

  const report = await testRunner.runTestsAndPrint(tests);

  expect(report).toMatchObject<SemiRunnerReport>({
    totalTestFiles: data.amountOfTestFiles,
    totalTestFilesFailed: 1,
    totalTestFilesPassed: 0,
    totalTests: 2,
    totalTestsFailed: 1,
    totalTestsPassed: 1,
    totalEmptyTestFiles: 0,
    totalEmptyTests: 0,
  });
  expect(removeANSIColorStyle(logUpdate.stder)).toMatchSnapshot();
});

it("should print report for 1 test file, 2 test clausure and 1 failed function and 1 passed function", async () => {
  const data: TestFileGeneratorInfo = {
    amountOfTestFiles: 1,
    testFunctionsReport: [
      {
        pass: false,
        message: buildReportMessage("expected: hi\n", "received: hi!"),
      },
      { pass: true },
    ],
    amountOfTests: 2,
  };

  const tests = generateTestFile(data);

  const report = await testRunner.runTestsAndPrint(tests);

  expect(report).toMatchObject<SemiRunnerReport>({
    totalTestFiles: data.amountOfTestFiles,
    totalTestFilesFailed: 1,
    totalTestFilesPassed: 0,
    totalTests: 4,
    totalTestsFailed: 2,
    totalTestsPassed: 2,
    totalEmptyTestFiles: 0,
    totalEmptyTests: 0,
  });
  expect(removeANSIColorStyle(logUpdate.stder)).toMatchSnapshot();
});

it("should print report for 2 test file, 2 test clausure and 1 failed function and 1 passed function", async () => {
  const data: TestFileGeneratorInfo = {
    amountOfTestFiles: 2,
    testFunctionsReport: [
      {
        pass: false,
        message: buildReportMessage("expected: hi\n", "received: hi!"),
      },
      { pass: true },
    ],
    amountOfTests: 2,
  };

  const tests = generateTestFile(data);

  const report = await testRunner.runTestsAndPrint(tests);

  expect(report).toMatchObject<SemiRunnerReport>({
    totalTestFiles: data.amountOfTestFiles,
    totalTestFilesFailed: 2,
    totalTestFilesPassed: 0,
    totalTests: 8,
    totalTestsFailed: 4,
    totalTestsPassed: 4,
    totalEmptyTestFiles: 0,
    totalEmptyTests: 0,
  });
  expect(removeANSIColorStyle(logUpdate.stder)).toMatchSnapshot();
});

it("should print report for 1 test file, 2 test clausure and 1 failed function and 1 passed function", async () => {
  const tests: TestFile[] = [
    {
      path: testFileNames[0],
      isEmpty: false,
      groups: [
        {
          name: "group name",
          tests: [
            {
              name: testNames[0],
              testsFunctions: [() => Promise.resolve({ pass: true })],
            },
            {
              name: testNames[1],
              testsFunctions: [
                () =>
                  Promise.resolve({
                    pass: false,
                    message: buildReportMessage("expected: hi\n", "received: hi!"),
                  }),
              ],
            },
          ],
        },
      ],
    },
  ];

  const report = await testRunner.runTestsAndPrint(tests);

  expect(report).toMatchObject<SemiRunnerReport>({
    totalTestFiles: 1,
    totalTestFilesFailed: 1,
    totalTestFilesPassed: 0,
    totalTests: 2,
    totalTestsFailed: 1,
    totalTestsPassed: 1,
    totalEmptyTestFiles: 0,
    totalEmptyTests: 0,
  });
  expect(removeANSIColorStyle(logUpdate.stder)).toMatchSnapshot();
});

it("should print for a empty test file name", async () => {
  const tests: TestFile[] = [
    {
      path: testFileNames[0],
      isEmpty: false,
      groups: [
        {
          name: "",
          tests: [
            {
              name: "",
              testsFunctions: [() => Promise.resolve({ pass: true })],
            },
          ],
        },
      ],
    },
  ];

  const report = await testRunner.runTestsAndPrint(tests);

  expect(report).toMatchObject<SemiRunnerReport>({
    totalTestFiles: 1,
    totalTestFilesFailed: 0,
    totalTestFilesPassed: 1,
    totalTests: 1,
    totalTestsFailed: 0,
    totalTestsPassed: 1,
    totalEmptyTestFiles: 0,
    totalEmptyTests: 0,
  });
  expect(removeANSIColorStyle(logUpdate.stder)).toMatchSnapshot();
});

it("should print tests for a subgroup", async () => {
  const tests: TestFile[] = [
    {
      path: testFileNames[0],
      isEmpty: false,
      groups: [
        {
          name: "",
          tests: [],
          subGroups: [
            {
              name: "subgroup",
              tests: [
                {
                  name: "test",
                  testsFunctions: [() => Promise.resolve({ pass: true })],
                },
              ],
            },
          ],
        },
      ],
    },
  ];

  const report = await testRunner.runTestsAndPrint(tests);

  expect(report).toMatchObject<SemiRunnerReport>({
    totalTestFiles: 1,
    totalTestFilesFailed: 0,
    totalTestFilesPassed: 1,
    totalTests: 1,
    totalTestsFailed: 0,
    totalTestsPassed: 1,
    totalEmptyTestFiles: 0,
    totalEmptyTests: 0,
  });
  expect(removeANSIColorStyle(logUpdate.stder)).toMatchSnapshot();
});

it("should print subtest for a subgroup", async () => {
  const tests: TestFile[] = [
    {
      path: testFileNames[0],
      isEmpty: false,
      groups: [
        {
          name: "",
          tests: [],
          subGroups: [
            {
              name: "subgroup",
              tests: [
                {
                  name: "test",
                  testsFunctions: [],
                  subTests: [
                    {
                      name: "subtest",
                      testsFunctions: [() => Promise.resolve({ pass: true })],
                    },
                  ],
                },
              ],
            },
          ],
        },
      ],
    },
  ];

  const report = await testRunner.runTestsAndPrint(tests);

  expect(report).toMatchObject<SemiRunnerReport>({
    totalTestFiles: 1,
    totalTestFilesFailed: 0,
    totalTestFilesPassed: 1,
    totalTests: 2,
    totalTestsFailed: 0,
    totalTestsPassed: 1,
    totalEmptyTestFiles: 0,
    totalEmptyTests: 1,
  });
  expect(removeANSIColorStyle(logUpdate.stder)).toMatchSnapshot();
});

it("should print empty test file", async () => {
  const tests: TestFile[] = [
    {
      path: testFileNames[0],
      isEmpty: true,
      groups: [],
    },
  ];

  const report = await testRunner.runTestsAndPrint(tests);

  expect(report).toMatchObject<SemiRunnerReport>({
    totalTestFiles: 1,
    totalTestFilesFailed: 0,
    totalTestFilesPassed: 0,
    totalTests: 0,
    totalTestsFailed: 0,
    totalTestsPassed: 0,
    totalEmptyTestFiles: 1,
    totalEmptyTests: 0,
  });
  expect(removeANSIColorStyle(logUpdate.stder)).toMatchSnapshot();
});

it("should print empty test file", async () => {
  const tests: TestFile[] = [
    {
      path: testFileNames[0],
      isEmpty: false,
      groups: [
        {
          name: "",
          tests: [],
          subGroups: [
            {
              name: "subgroup",
              tests: [
                {
                  name: "test",
                  testsFunctions: [],
                  subTests: [
                    {
                      name: "subtest",
                      testsFunctions: [() => Promise.resolve({ pass: true })],
                    },
                  ],
                },
              ],
            },
          ],
        },
      ],
    },
    {
      groups: [],
      isEmpty: true,
      path: testFileNames[1],
    },
  ];

  const report = await testRunner.runTestsAndPrint(tests);

  expect(report).toMatchObject<SemiRunnerReport>({
    totalTestFiles: 2,
    totalTestFilesFailed: 0,
    totalTestFilesPassed: 1,
    totalTests: 2,
    totalTestsFailed: 0,
    totalTestsPassed: 1,
    totalEmptyTestFiles: 1,
    totalEmptyTests: 1,
  });
  expect(removeANSIColorStyle(logUpdate.stder)).toMatchSnapshot();
});

it("should print report for 1 test file, 2 test clausure and 1 failed function and 1 passed function", async () => {
  const data: TestFileGeneratorInfo = {
    amountOfTestFiles: 1,
    testFunctionsReport: [
      {
        pass: false,
        message: buildReportMessage(diff({ a: 1 }, { a: 2 })),
      },
    ],
    amountOfTests: 1,
  };

  const tests = generateTestFile(data);

  const report = await testRunner.runTestsAndPrint(tests);

  expect(report).toMatchObject<SemiRunnerReport>({
    totalTestFiles: data.amountOfTestFiles,
    totalTestFilesFailed: 1,
    totalTestFilesPassed: 0,
    totalTests: 1,
    totalTestsFailed: 1,
    totalTestsPassed: 0,
    totalEmptyTestFiles: 0,
    totalEmptyTests: 0,
  });
  expect(removeANSIColorStyle(logUpdate.stder)).toMatchSnapshot();
});
