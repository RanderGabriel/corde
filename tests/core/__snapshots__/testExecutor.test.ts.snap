// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should print empty test file 1`] = `" EMPTY   /tests/file1.test.ts"`;

exports[`should print empty test file 2`] = `
"
 PASS   /tests/file1.test.ts   2ms
    ● test (empty)   
   ✔️  subtest   1ms EMPTY   /tests/file2.test.ts"
`;

exports[`should print for a empty test file name 1`] = `
"
 PASS   /tests/file1.test.ts   2ms
   ✔️  <empty test name>   "
`;

exports[`should print report for 1 test file, 1 test clausure and 1 failed function and 1 passed function 1`] = `
"
 FAIL   /tests/file1.test.ts   3ms
   ❌ group -> test case1   1ms
    
    expected: hireceived: hi!
"
`;

exports[`should print report for 1 test file, 1 test clausure and 1 failed function and 1 passed function 2`] = `
"
 FAIL   /tests/file1.test.ts   2ms
   ❌ group -> test case1   
    
    expected: hireceived: hi!
"
`;

exports[`should print report for 1 test file, 1 test clausure and 1 test function 1`] = `
"
 PASS   /tests/file1.test.ts   10ms
   ✔️  group -> test case1   3ms"
`;

exports[`should print report for 1 test file, 1 test clausure and 1 test function that fail 1`] = `
"
 FAIL   /tests/file1.test.ts   2ms
   ❌ group -> test case1   1ms
    
    expected: hireceived: hi!
"
`;

exports[`should print report for 1 test file, 1 test clausure and 2 test function 1`] = `
"
 PASS   /tests/file1.test.ts   4ms
   ✔️  group -> test case1   1ms"
`;

exports[`should print report for 1 test file, 2 test clausure and 1 failed function and 1 passed function 1`] = `
"
 FAIL   /tests/file1.test.ts   7ms
   ❌ group -> test case1   1ms
    
    expected: hireceived: hi!

   ❌ group -> test case2   3ms
    
    expected: hireceived: hi!
"
`;

exports[`should print report for 1 test file, 2 test clausure and 1 failed function and 1 passed function 2`] = `
"
 FAIL   /tests/file1.test.ts   3ms
   ✔️  group name -> test case1   1ms
   ❌ group name -> test case2   1ms
    
    expected: hireceived: hi!
"
`;

exports[`should print report for 1 test file, 2 test clausure and 1 failed function and 1 passed function 3`] = `
"
 FAIL   /tests/file1.test.ts   2ms
   ❌ group -> test case1   1ms
    
    - expect    - 1
    + received  + 1
    
      Object {
    -   \\"a\\": 1,
    +   \\"a\\": 2,
      }
"
`;

exports[`should print report for 2 test file, 1 test clausure and 1 test function 1`] = `
"
 PASS   /tests/file1.test.ts   2ms
   ✔️  group -> test case1   
 PASS   /tests/file2.test.ts   2ms
   ✔️  group -> test case1   1ms"
`;

exports[`should print report for 2 test file, 1 test clausure and 2 test function 1`] = `
"
 PASS   /tests/file1.test.ts   1ms
   ✔️  group -> test case1   1ms
 PASS   /tests/file2.test.ts   4ms
   ✔️  group -> test case1   1ms"
`;

exports[`should print report for 2 test file, 2 test clausure and 1 failed function and 1 passed function 1`] = `
"
 FAIL   /tests/file1.test.ts   5ms
   ❌ group -> test case1   
    
    expected: hireceived: hi!

   ❌ group -> test case2   1ms
    
    expected: hireceived: hi!

 FAIL   /tests/file2.test.ts   5ms
   ❌ group -> test case1   
    
    expected: hireceived: hi!

   ❌ group -> test case2   1ms
    
    expected: hireceived: hi!
"
`;

exports[`should print report for 2 test file, 2 test clausure and 2 test function 1`] = `
"
 PASS   /tests/file1.test.ts   3ms
   ✔️  group -> test case1   
   ✔️  group -> test case2   1ms
 PASS   /tests/file2.test.ts   4ms
   ✔️  group -> test case1   1ms
   ✔️  group -> test case2   1ms"
`;

exports[`should print subtest for a subgroup 1`] = `
"
 PASS   /tests/file1.test.ts   2ms
    ● test (empty)   
   ✔️  subtest   "
`;

exports[`should print tests for a subgroup 1`] = `
"
 PASS   /tests/file1.test.ts   1ms
   ✔️  test   "
`;
